src/backend/core/models.py

```diff
@@ -325,3 +388,225 @@ def clean_fields(self, exclude=None):
+
+        # Add owner role if user is owner
+        if is_owner:
+            assignable_roles.add(RoleChoices.OWNER)
+
+        # Remove current role as it can't be assigned
+        assignable_roles.discard(self.role)
+
+        return sorted(str(role) for role in assignable_roles)
+
+    def _get_abilities(self, resource: models.Model, user: User) -> Dict[str, bool]:
+        """
+        Compute all abilities for a user on a given resource.
+
+        This method aggregates all permissions and provides a comprehensive
+        dictionary of allowed actions for the resource.
+
+        Args:
+            resource: The resource to check abilities for
+            user: The user to compute abilities for
+
+        Returns:
+            Dictionary mapping ability names to boolean permissions
+        """
+
+        # Default abilities, with no rights
+        abilities = {
+            "destroy": False,
+            "update": False,
+            "partial_update": False,
+            "retrieve": False,
+            "set_role_to": [],
+        }
+
+        # Early returns for unauthenticated users
+        if not user.is_authenticated:
+            return abilities
+
+        roles = get_resource_roles(resource, user)
+
+        # if no roles on the resource early return with no abilities
+        if not roles:
+            return abilities
+
+        abilities["retrieve"] = True
+
+        is_admin = RoleChoices.ADMIN in roles
+        is_owner = RoleChoices.OWNER in roles
+
+        # Restrict to read-only if insufficient privileges
+        if not (is_admin or is_owner) or (
+            self.role == RoleChoices.OWNER and not is_owner
+        ):
+            return abilities
+
+        # Allow owner deletion only if not the last owner
+        if self.role == RoleChoices.OWNER:
+            if is_owner:
+                owner_count = resource.accesses.filter(role=RoleChoices.OWNER).count()
+                abilities["destroy"] = owner_count > 1
+        else:
+            abilities["destroy"] = is_owner or is_admin
+
+        # Determine role modification permissions
+        can_modify_roles = abilities["destroy"] and not (
+            self.role == RoleChoices.OWNER and not is_owner
+        )
+        assignable_roles = self._get_assignable_roles(can_modify_roles, is_owner)
+
+        abilities.update(
+            {
+                "update": bool(assignable_roles),
+                "partial_update": bool(assignable_roles),
+                "set_role_to": assignable_roles,
+            }
+        )
+
+        return abilities
+
+
+class Recording(BaseModel):
+    """Model for recordings that take place in a room"""
+
+    room = models.ForeignKey(
+        Room,
+        on_delete=models.CASCADE,
+        related_name="recordings",
+        verbose_name=_("Room"),
+    )
+    status = models.CharField(
+        max_length=20,
+        choices=RecordingStatusChoices.choices,
+        default=RecordingStatusChoices.INITIATED,
+    )
```

Comment:
Should I add db constraint to avoid transitioning from a state to another (e.g. `Active -> Initiated`). 

Normal flow would look like :
`Initiated -> Active -> Stopped -> Saved`

Reply:
Not sure a DB constraint is needed, but I would definitely add this expected workflow in the model docstring ;)

Reply:
true